### PR TITLE
[android] Android Auto support. `MapFragment.java` refactoring

### PR DESCRIPTION
**Pull request by AndrewShkrob**
I'd like to start integrating this feature with master.
This is the first step.
All common logic for the mobile app and AA is moved into `Map.java`.
Here you can check how `Map.java` will be used in the AA: #3806 `android/flavors/car/src/app/organicmaps/car/SurfaceRenderer.java`

### CODE REVIEW

File: android/jni/app/organicmaps/Map.cpp [hunk 1]
```diff
@@ -22,147 +22,159 @@ void OnRenderingInitializationFinished(std::shared_ptr<jobject> const & listener
 
 extern "C"
 {
-using namespace storage;
+JNIEXPORT jboolean JNICALL
+Java_app_organicmaps_Map_nativeCreateEngine(JNIEnv * env, jclass clazz,
+                                                        jobject surface, jint density,
```
Comment (biodranik) [inline android/jni/app/organicmaps/Map.cpp:27]: 1. It doesn't look aligned.
2. Please avoid declaring unused function parameters like clazz, or comment them out.

File: android/src/app/organicmaps/MapFragment.java [hunk 2]
```diff
@@ -1,251 +1,75 @@
...
-    nativeSetupWidget(WIDGET_SCALE_FPS_LABEL,
-                      UiUtils.dimen(context, R.dimen.margin_base),
-                      UiUtils.dimen(context, R.dimen.margin_base),
-                      ANCHOR_LEFT_TOP);
-
-    setupCompass(mCurrentCompassOffsetY, mCurrentCompassOffsetX, false);
-  }
-
-  /**
-   * Moves the map compass using the given offsets.
```
Comment (biodranik) [inline android/src/app/organicmaps/MapFragment.java:97]: Did you intentionally remove comments here and in other places??

**Review by biodranik — COMMENTED** [review]
This change looks a bit dangerous. Did you test it in all modes, including deep linking? You can try our test page here: https://omaps.app/test

File: android/src/app/organicmaps/MapFragment.java (hunk 2 above)
Comment (AndrewShkrob) [reply android/src/app/organicmaps/MapFragment.java:97]: I moved them to `Map.java`

**Comment by AndrewShkrob** [comment]
>This change looks a bit dangerous. Did you test it in all modes, including deep linking? You can try our test page here: https://omaps.app/test

I did some basic tests. Open the app, close, rotate, build a route, search and so on. It works like before. No crashes, no lags. I also checked for memory leaks and didn't find anything.

Regarding https://omaps.app/test:
It works like the version from google play:
* Google URLs are not opening (probably because I have Google Maps installed)
* OSM links 1 and 3 are not opening (both on google play and this branch versions)

So, nothing changed

**Comment by biodranik** [comment]
Thanks! @rtsisyk can you please review, and also check why mentioned OSM links are not opening?

File: android/jni/app/organicmaps/Map.cpp [hunk 3]
```diff
@@ -22,147 +22,156 @@ void OnRenderingInitializationFinished(std::shared_ptr<jobject> const & listener
...
+                                            jint appVersionCode)
+{
+  return g_framework->CreateDrapeEngine(env, surface, density, firstLaunch, isLaunchByDeepLink,
+                                        base::asserted_cast<uint32_t>(appVersionCode));
+}
 
-JNIEXPORT void JNICALL
-Java_app_organicmaps_MapFragment_nativeCompassUpdated(JNIEnv * env, jclass clazz, jdouble north, jboolean forceRedraw)
+JNIEXPORT jboolean JNICALL
+Java_app_organicmaps_Map_nativeIsEngineCreated(JNIEnv * env, jclass)
```
Comment (biodranik) [inline android/jni/app/organicmaps/Map.cpp:37]:
```suggestion
Java_app_organicmaps_Map_nativeIsEngineCreated(JNIEnv *, jclass)
```

File: android/jni/app/organicmaps/Map.cpp [hunk 4]
```diff
@@ -22,147 +22,156 @@ void OnRenderingInitializationFinished(std::shared_ptr<jobject> const & listener
...
+JNIEXPORT jboolean JNICALL
+Java_app_organicmaps_Map_nativeShowMapForUrl(JNIEnv * env, jclass, jstring url)
+{
+  return g_framework->ShowMapForURL(jni::ToNativeString(env, url));
 }
 
 JNIEXPORT void JNICALL
-Java_app_organicmaps_MapFragment_nativeStorageConnected(JNIEnv * env, jclass clazz)
+Java_app_organicmaps_Map_nativeSetRenderingInitializationFinishedListener(
+  JNIEnv * env, jclass, jobject listener)
```
Comment (biodranik) [inline android/jni/app/organicmaps/Map.cpp:50]:
```suggestion
  JNIEnv *, jclass, jobject listener)
```

File: android/jni/app/organicmaps/Map.cpp [hunk 5]
```diff
@@ -22,147 +22,156 @@ void OnRenderingInitializationFinished(std::shared_ptr<jobject> const & listener
...
+Java_app_organicmaps_Map_nativeAttachSurface(JNIEnv * env, jclass, jobject surface)
 {
-  android::Platform::Instance().OnExternalStorageStatusChanged(false);
-  g_framework->RemoveLocalMaps();
+  return g_framework->AttachSurface(env, surface);
 }
 
 JNIEXPORT void JNICALL
-Java_app_organicmaps_MapFragment_nativeScalePlus(JNIEnv * env, jclass clazz)
+Java_app_organicmaps_Map_nativeDetachSurface(JNIEnv * env, jclass, jboolean destroySurface)
```
Comment (biodranik) [inline android/jni/app/organicmaps/Map.cpp:70]:
```suggestion
Java_app_organicmaps_Map_nativeDetachSurface(JNIEnv *, jclass, jboolean destroySurface)
```

Please clean up other cases too.

File: android/src/app/organicmaps/Map.java [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+   * Moves the map compass using the given offsets.
+   *
+   * @param context     Context.
+   * @param offsetX     Pixel offset from the top. -1 to keep the previous value.
+   * @param offsetY     Pixel offset from the right.  -1 to keep the previous value.
+   * @param forceRedraw True to force the compass to redraw
+   */
+  public void setupCompass(final Context context, int offsetX, int offsetY, boolean forceRedraw)
+  {
+    int x = offsetX < 0 ? mCurrentCompassOffsetX : offsetX;
```
Comment (biodranik) [inline android/src/app/organicmaps/Map.java:86]: Is it hard to mark all const java vars with `final` keyword?

File: android/src/app/organicmaps/Map.java [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+    mBottomWidgetOffsetX = x;
+    mBottomWidgetOffsetY = y;
+  }
+
+  public void onSurfaceCreated(final Context context, final Surface surface, Rect surfaceFrame, int surfaceDpi)
+  {
+    if (nativeIsEngineCreated())
+    {
+      nativeDetachSurface(true);
+    }
```
Comment (biodranik) [inline android/src/app/organicmaps/Map.java:128]:
```suggestion
    if (nativeIsEngineCreated())
      nativeDetachSurface(true);
```

We prefer smaller code when possible.

File: android/src/app/organicmaps/Map.java [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+      nativeResumeSurfaceRendering();
+      return;
+    }
+
+    mRequireResize = false;
+    setupWidgets(context, surfaceFrame.width(), surfaceFrame.height());
+
+    final boolean firstStart = LocationHelper.INSTANCE.isInFirstRun();
+    if (!nativeCreateEngine(surface, surfaceDpi, firstStart, mLaunchByDeepLink,
+        BuildConfig.VERSION_CODE))
```
Comment (biodranik) [inline android/src/app/organicmaps/Map.java:157]: It can be in one line.

File: android/src/app/organicmaps/Map.java [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+    {
+      if (mCallbackUnsupported != null)
+        mCallbackUnsupported.report();
+      return;
+    }
+
+    if (firstStart)
+    {
+      UiThread.runLater(LocationHelper.INSTANCE::onExitFromFirstRun);
+    }
```
Comment (biodranik) [inline android/src/app/organicmaps/Map.java:167]:
```suggestion
    if (firstStart)
      UiThread.runLater(LocationHelper.INSTANCE::onExitFromFirstRun);
```

File: android/src/app/organicmaps/MapFragment.java [hunk 10]
```diff
@@ -1,251 +1,75 @@
...
-    mSurfaceCreated = true;
-    mSurfaceAttached = true;
-    nativeResumeSurfaceRendering();
-    if (mMapRenderingListener != null)
-      mMapRenderingListener.onRenderingCreated();
+    mMap.onSurfaceCreated(
+        requireContext(),
+        surfaceHolder.getSurface(),
+        surfaceHolder.getSurfaceFrame(),
+        densityDpi);
```
Comment (biodranik) [inline android/src/app/organicmaps/MapFragment.java:62]:
```suggestion
    mMap.onSurfaceCreated(requireContext(), surfaceHolder.getSurface(), surfaceHolder.getSurfaceFrame(), densityDpi);
```

File: android/src/app/organicmaps/MapFragment.java [hunk 11]
```diff
@@ -1,251 +1,75 @@
...
-
-    final Surface surface = surfaceHolder.getSurface();
-    nativeSurfaceChanged(surface, width, height);
-
-    mRequireResize = false;
-    setupWidgets(width, height);
-    nativeApplyWidgets();
-    if (mMapRenderingListener != null)
-      mMapRenderingListener.onRenderingRestored();
+    mMap.onSurfaceChanged(
```
Comment (biodranik) [inline android/src/app/organicmaps/MapFragment.java:68]: Make one line?

File: android/src/app/organicmaps/MapFragment.java [hunk 12]
```diff
@@ -255,170 +79,119 @@ public void surfaceDestroyed(SurfaceHolder surfaceHolder)
...
-      nativeResumeSurfaceRendering();
+    mMap.onResume();
   }
 
   @Override
   public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState)
   {
     View view = inflater.inflate(R.layout.fragment_map, container, false);
-    mSurfaceView = view.findViewById(R.id.map_surfaceview);
+    SurfaceView mSurfaceView = view.findViewById(R.id.map_surfaceview);
```
Comment (biodranik) [inline android/src/app/organicmaps/MapFragment.java:143]: final?

**Review by biodranik — APPROVED** [review]
It works for me. @rtsisyk please review

**Comment by rtsisyk** [comment]
> It works for me. @rtsisyk please review

I will take a look.

**Review by rtsisyk — APPROVED** [review]
Good refactoring. 
Thanks for contributing!

**Pull request merged**